File: src/main/java/model/moves/FlyingKingMove.java
package model.moves;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import model.Cell;
import model.Checkers;
import model.State;

public class FlyingKingMove extends Moves {

    public FlyingKingMove(Checkers c) {
        this.checkers = c;
    }

    @Override
    public List<List<Cell>> possibleMoves(Cell currentCell) {
        int crow = currentCell.x(), ccol = currentCell.y();
        int[][] pos = {{-1, -1}, {-1, 1}, {1, -1}, {1, 1}};
        List<List<Cell>> moves = new ArrayList<>();
        for (int[] po : pos) {
            int diffRow = po[0], diffCol = po[1], i = 1;
            do {
                Cell destinationCell = new Cell(crow + i * diffRow, ccol + i * diffCol);
                if (!checkers.onBoard(destinationCell) || checkers.getState(destinationCell) != State.EMPTY) {
                    break;
                } else {
                    List<Cell> temp = new ArrayList<>();
                    temp.add(destinationCell);
                    moves.add(temp);
                }
                i++;
            } while (true);
        }
        moves.forEach(l -> Collections.reverse(l));
        return moves;
    }
}

File: docs/scripts/moves/FlyingKingMove.js
import Cell from '../Cell.js';
import State from '../State.js';

export default class FlyingKingMove {
    constructor(c) {
        this.checkers = c;
    }
    possibleMoves({ x: crow, y: ccol }) {
        let pos = [[-1, -1], [-1, 1], [1, -1], [1, 1]], moves = [];
        for (let [diffRow, diffCol] of pos) {
            for (let i = 1; ; i++) {
                let destinationCell = new Cell(crow + i * diffRow, ccol + i * diffCol);
                if (!this.checkers.onBoard(destinationCell) || this.checkers.getState(destinationCell) !== State.EMPTY) {
                    break;
                } else {
                    moves.push([destinationCell]);
                }
            }
        }
        return moves;
    }
}